### PR TITLE
[Filebeat][New Module] Threat intel module for filebeat

### PR DESCRIPTION
**Pull request by P1llus**
## What does this PR do?

This PR adds a new module for threat intel, using each TI source as a fileset.

## Why is it important?

Adds possibility to ingest Threat Intel data to be used for security usecases

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

Resolves https://github.com/elastic/beats/issues/23406

### CODE REVIEW

File: filebeat/docs/fields.asciidoc
Comment (FrankHassanabad) [inline filebeat/docs/fields.asciidoc:151084]: nit: Reporter instead of repoter?

File: x-pack/filebeat/module/threatintel/abuseurl/_meta/fields.yml
Comment (FrankHassanabad) [inline x-pack/filebeat/module/threatintel/abuseurl/_meta/fields.yml:34]: nit: Another repoter instead of reporter I think.

File: x-pack/filebeat/module/threatintel/abuseurl/test/abusechurl.ndjson.log-expected.json
Comment (FrankHassanabad) [inline x-pack/filebeat/module/threatintel/abuseurl/test/abusechurl.ndjson.log-expected.json:25]: nit: even though this historical record is no longer malicious I would just be very very careful generally using potential maleware examples in source code for tests as someone could accidentally activate them or the tests could somehow at a later point in time.

**Comment by P1llus** [comment]
@FrankHassanabad and @leehinman Added in your comments, feel free to let me know if its anything else we should change as well.

**Comment by leehinman** [comment]
Looks like `larted` is still getting stored as a string:

```
"threatintel.larted": "true",
```

Might be a good idea to use a convert to store it as a boolean so it doesn't have to be converted when used.

**Comment by shimonmodi** [comment]
@P1llus - I have a threat intel ECS RFC in progress right now, and want to apply those ECS fields to this module. the mapping is available here [https://github.com/elastic/security-team/issues/177](https://github.com/elastic/security-team/issues/177) . Whats the best way to reconcile the mappings you have used with the one's in the RFC?

**Comment by P1llus** [comment]
@shimonmodi If we want to support the ECS RFC for threat intel whenever we release the module then I don't mind converting it.

I was just unsure if this RFC was complete enough yet to use, if it is then I can convert this module before we proceed :)

**Comment by P1llus** [comment]
Sorry about that @leehinman . I only updated the field mapping and didn't actually convert it. Added the update now. Hopefully everything else is alright? :)

**Comment by P1llus** [comment]
Let's not merge it yet though, but good to know if its "LGTM" or not!

**Comment by leehinman** [comment]
> Sorry about that @leehinman . I only updated the field mapping and didn't actually convert it. Added the update now. Hopefully everything else is alright? :)

LGTM

**Review by leehinman — APPROVED** [review]

**Comment by shimonmodi** [comment]
@P1llus - perfect, lets chat about it on Wednesday when we meet. My goal is to validate that the proposed ECS fields work for this feed and identify gaps.

**Comment by P1llus** [comment]
I think I will leave this PR as draft for now, that way we can modify it a bit more while testing and discussing with different teams. There might be some breaking changes and temp content in here, so I think for any new content we can hold off new reviews for now.

**Comment by P1llus** [comment]
Requires https://github.com/elastic/beats/pull/22320 to finalize

**Comment by P1llus** [comment]
Newest commit does a huge rewrite of the whole module to fit the new TI ECS fields, more test data usecases, moving over to httpjson v2 input with cursor support, updating the docs and the config templates.

File: x-pack/filebeat/module/threatintel/abusemalware/_meta/fields.yml
Comment (peasead) [inline x-pack/filebeat/module/threatintel/abusemalware/_meta/fields.yml:49]:
```suggestion
```
Can we move `ssdeep` and `tlsh` to `threat.indicator.file.hash.{ssdeep,tlsh}`? I know they're not part of the `hash.*` fieldset, but we can add them and this is where they'd end up.

For `impash`, can we move that to `threat.indicator.file.pe.imphash`? It's part of the PE fieldset now.

**Comment by andrewkroh** [comment]
Regarding testing this, there are three types of tests that would be useful, but not all are required.

1. Pipeline tests. Capture some events from the respective services and write them to file. Then push them through the ingest node pipeline to ensure it does the processing we expect. This is what we do for almost every module today.
2. Unit test the httpjson configuration against a mock service. These are great for quickly detecting breakages during Filebeat development, but won't catch issues if the service diverges from the mock. I'd say these are the lowest priority.
3. Integration test the module with the real service. We don't have anything like this today for Filebeat modules, but there might be some prior art for Metricbeat modules that work with cloud service APIs. We'd load any secrets needed by the tests (like api keys) into Vault then CI would grab those in order to run the tests. We wouldn't want these tests to run by default for developers.

**Comment by P1llus** [comment]
> Regarding testing this, there are three types of tests that would be useful, but not all are required.
> 
> 1. Pipeline tests. Capture some events from the respective services and write them to file. Then push them through the ingest node pipeline to ensure it does the processing we expect. This is what we do for almost every module today.
> 2. Unit test the httpjson configuration against a mock service. These are great for quickly detecting breakages during Filebeat development, but won't catch issues if the service diverges from the mock. I'd say these are the lowest priority.
> 3. Integration test the module with the real service. We don't have anything like this today for Filebeat modules, but there might be some prior art for Metricbeat modules that work with cloud service APIs. We'd load any secrets needed by the tests (like api keys) into Vault then CI would grab those in order to run the tests. We wouldn't want these tests to run by default for developers.

I think it would be great to have 2 and 3 as well, 1 has already been added for each fileset itself. I do think httpjson input has its unittests but we can always extend it.
For 3 however it is something we can surely look into, but might need to be added outside of this PR.

Just to mentions, we have access to all of these sources and both me and internal teams are testing them as we speak.

**Comment by P1llus** [comment]
The current status of this PR is that it should be in a mostly complete state, but its currently missing the TI ECS fields, currently some dicussions are ongoing on how we want to handle that.

**Comment by P1llus** [comment]
The module is now ready for review!

**Comment by P1llus** [comment]
jenkins run tests please

**Comment by P1llus** [comment]
jenkins run tests please

**Comment by P1llus** [comment]
Added some fixes above for general feedback received outside of the PR. All planned changes has been applied.

File: filebeat/docs/modules/threatintel.asciidoc
Comment (andrewkroh) [inline filebeat/docs/modules/threatintel.asciidoc:15]: I think it would be good to provide context around how this data can be utilized with the Elastic Stack (threat indicator rules and maybe enrichment processors?).

File: x-pack/filebeat/module/threatintel/abusemalware/manifest.yml
Comment (andrewkroh) [inline x-pack/filebeat/module/threatintel/abusemalware/manifest.yml:10]: Please ensure parameters without defaults are protected with `{{ if .X }}....{{ end }}` in the templates. I'm pretty sure our renderer is setup to error on undefined variables.

File: x-pack/filebeat/module/threatintel/_meta/fields.yml
Comment (andrewkroh) [inline x-pack/filebeat/module/threatintel/_meta/fields.yml:1]: IIRC there's a `release: beta` tag for fields.yml.

**Review by andrewkroh — COMMENTED** [review]
I reviewed the module, but not the data mappings. It looks like it's in good shape with the exception of a few really minor comments below. Before merging please create a ticket to track the work necessary to add automated integration tests for the httpjson side of the filesets.

File: x-pack/filebeat/module/threatintel/_meta/docs.asciidoc
Comment (adriansr) [inline x-pack/filebeat/module/threatintel/_meta/docs.asciidoc:10]: nit

```suggestion
This module is a collection of different threat intelligence sources. The ingested data is meant to be used with [Indicator Match rules]https://www.elastic.co/guide/en/security/7.11/rules-ui-create.html#create-indicator-rule, but is also
```

**Review by adriansr — APPROVED** [review]
LGTM

**Pull request merged**
